### DISCUSSION TITLE
Bitstream manipulation for LUTs

### DISCUSSION THREAD

**Discussion by gabrielcobostello** [Xilinx/RapidWright#295] [Q&A] [answered]
Greetings!

I wanted to ask whether it is possible to replicate the bitstream manipulation example in the documentation but using a LUT as the target.
I have basically written the same code as in the example, the only difference being that my cell is a LUT4, but for some reason the configArray does not seem to get updated. Is this due to rapidwright not knowing the mapping of the LUTs in the configuration memory, or am I doing something wrong ?
```
cell.getProperty("INIT").setValue("16'h1111");  //Where cell is the LUT4.
configArray.updateUserStateBits(cell);
```
The change does apear in the dcp when loaded into vivado.

Thx for any clarification.
Best regards,
Gabriel.

**Comment by clavin-xlnx** [comment, marked as answer]
Hi Gabriel,

Since LUTs do not contain "User State Bits" (a user design does not change them), they do not get updated by the `ConfigArray.updateUserStateBits()`.  We can only provide functionality that is clearly documented and, unfortunately, this is not the case for INIT values on LUTs.  The complete list of bits that we can update can be generated for a design by calling [`write_bitstream -logic_location_file`](https://www.xilinx.com/support/documentation/sw_manuals/xilinx2021_1/ug835-vivado-tcl-commands.pdf#nameddest=xwrite_bitstream).  

Sorry we can't offer more on this topic.

Chris

**Reply by gabrielcobostello** [reply]
Thx Chris,
Yes I'm familiar with the .ll file, its unfortunate that those are the only documented bits.
We have LUTs mapped for series7 fpgas, and where hoping to skip the tedious mapping process for ultrascale and ultrascale+  fpgas with rapidwright, but it seems this wont be possible.

Thx for the clarification, and I would suggest specifying this regarding the .ll file on the disclaimer section of the rapidwright bitstream manipulation documentation, as in my case it was a bit confusing finding out what could be or could not be done.

Best regards,
Gabriel.

**Reply by clavin-xlnx** [reply]
Thanks Gabriel for the feedback.   

I do have a reference to this issue in the [API Javadoc](http://www.rapidwright.io/javadoc/com/xilinx/rapidwright/bitstream/ConfigArray.html#updateUserStateBits-com.xilinx.rapidwright.design.Cell...-), however, I have added [a similar note in the documentation](http://www.rapidwright.io/docs/Bitstream_Manipulation.html#example-usages-modify-user-state-bits).

Chris
